014-single-codebase.md

```diff
@@ -0,0 +1,68 @@
+
+
+
+## Proposal
+
+We should consolidate our backend services (Northstar, Rogue, and Chompy) into **a single monolithic Laravel application.**
+
+- This will **simplify architectural decisions** & enable us to iterate faster. Instead of sorting new functionality into strict bounded contexts up front, we'll be able to simply place new entities into a single backend context. Architectural decisions will be less binding, as it's easier to refactor within a single codebase than across network boundaries.
+
+- It will also **simplify adding behavior that straddles contexts**, like clubs, by replacing network requests with function calls. We'll also be able to ensure better consistency by using more transactional validation.
+
+- Keeping more of our business entities in the same domain allows **easier integration testing**, since we can seed (and otherwise directly reference) data that would currently live across API boundaries, like user IDs & clubs.
+
+- We have less applications to keep up-to-date with security & framework upgrades!
+
+For services that may not make sense to fold into the monolith, like [GraphQL](https://github.com/DoSomething/graphql/) or [Gambit](https://github.com/DoSomething/gambit/), we should still collaborate on them [**as part of a single monorepo**](https://docs.google.com/document/d/1a3e0Wpv9qeRscspdbN4wVR4Chz5oJFb9HH69x2pngSg/edit), side-by-side with our consolidated backend application. This allows us to retain the technical benefits we've gotten from separating out these services, while collaborating in one place.
+
+
+
+## Drawbacks/Alternatives
+
+#### Okay, but on the flip side – why _did_ we move from a monolith to (micro)services?
+
+Our Drupal application had a lot of intertwined modules that could be hard to follow, due to Drupal 6's "hooks" based architecture. We wanted to use service boundaries to force a "separation of concerns" that we thought would keep our codebase tidier.
+
+At the time, microservices were the architecture du jour & we followed the success stories of teams many times our size, from Spotify to Google, that promised that smaller services would increase velocity and allow teams to ship faster. At first, we had roughly one team of developers per application ("Rocket" for Phoenix & Northstar, "Bleed" for Rogue, and "O'Doyle" for Gambit & Blink) and it really did increase velocity! (That velocity slowed as a we began to build more cross-cutting features later on.)
+
+We also found that splitting our monolithic application into services allowed us to choose "the best tool for the job", such as using Node.js for our chatbot. (We love PHP at Do Something, but we know it's not appropriate for every task).
```

Comment:
I can't help myself from commenting here that this line makes me daydream of an all JS stack (using Node.js instead of Laravel).. but definitely doesn't make sense as a first pass for Monolith v1

Reply:
Yeah, being able to use the same language & code across the whole stack is just so incredibly compelling. I'm not sure anything compares with Laravel for "out of the box" productivity in the JavaScript ecosystem, though. (Although I am very eagerly following frameworks like [RedwoodJS](https://redwoodjs.com) that seem to be filling that niche!!)

Reply:
Alternatively, we could go in the direction of a full Laravel stack, utilizing [Blade components](https://laravel.com/docs/8.x/blade#components) for all our component-ized code needs, [Laravel Livewire](https://laravel-livewire.com/) for no-page-refresh updates (get that SPA-feel) and some sprinkling of [AlpineJS](https://github.com/alpinejs/alpine) where needed. And we get server-side rendering (pages and components) cause they're Blade views that get cached along with the great "out of the box" productivity and security that comes with the Laravel ecosystem. 😈  

I've used a bit of this on my own sites and it's pretty rad; down for whatever the team prefers, just wanna make sure we're keeping options open 😚

Reply:
It does look incredibly slick, and makes a ton of sense for a smaller team like ours.

From a practical perspective, I'm wary of anything that'd involve too much rewriting (either from Laravel to Node.js, or from React to Livewire/Alpine), since that leaves us potentially "spinning our wheels" for a long time. If we were to consider something like that, I think we'd need a good plan on how we'd manage that transition effectively.

Reply:
Yep totally agree @DFurnes. I mostly just brought it up so as to keep the TALL (Tailwind, AlpineJS, Livewire & Laravel) Stack in consideration and if it could serve our purposes for some apps (or depending on how it comes together as a single repo), especially if some of our stack is already heavy in some parts with Laravel.

I think for Phoenix and the UI for the main website it makes sense to continue down the NodeJS path and more specifically ReactJS & NextJS, but I have thought in passing that the TALL stack could work great for our more backend heavy functionality like user "Activity" (what Rogue handles). I think the Rogue Admin UI in particular, would be a solid candidate to switch it to componentize it with Blade components and use AlpineJS, Livewire and Tailwind for styling.

But ➕ to needing to have a good plan and discuss how to manage it and even _if_ it would makes sense to go that route! Just sharing thoughts for now 👯‍♂️